PayPal/src/main/java/com/braintreepayments/api/PayPalLauncher.kt

```diff
@@ -0,0 +1,94 @@
+ */
+class PayPalLauncher internal constructor(private val browserSwitchClient: BrowserSwitchClient) {
+    /**
+     * Used to launch the PayPal flow in a web browser and deliver results to your Activity
+     */
+    constructor() : this(BrowserSwitchClient())
+
+    /**
+     * Launches the PayPal flow by switching to a web browser for user authentication
+     *
+     * @param activity the Android Activity from which you will launch the web browser
+     * @param paymentAuthRequest a [PayPalPaymentAuthRequest.ReadyToLaunch] received from
+     * calling [PayPalClient.createPaymentAuthRequest]
+     * @return [PayPalPendingRequest] a [PayPalPendingRequest.Started] should be stored
+     * to complete the flow upon return to app in
+     * [PayPalLauncher.handleReturnToAppFromBrowser],
+     * or a [PayPalPendingRequest.Failure] with an error if the PayPal flow was unable to be
+     * launched in a browser.
+     */
+    fun launch(
+        activity: ComponentActivity,
+        paymentAuthRequest: PayPalPaymentAuthRequest.ReadyToLaunch
+    ): PayPalPendingRequest {
+        try {
+            assertCanPerformBrowserSwitch(activity, paymentAuthRequest.requestParams)
+        } catch (browserSwitchException: BrowserSwitchException) {
+            val manifestInvalidError = createBrowserSwitchError(browserSwitchException)
+            return PayPalPendingRequest.Failure(manifestInvalidError)
+        }
+        val request = browserSwitchClient.start(
+            activity,
+            paymentAuthRequest.requestParams.browserSwitchOptions
+        )
+        return when (request) {
+            is BrowserSwitchPendingRequest.Failure -> PayPalPendingRequest.Failure(request.cause)
+            is BrowserSwitchPendingRequest.Started -> PayPalPendingRequest.Started(request)
+        }
+    }
+
+    /**
+     * Captures and delivers the result of a the browser-based PayPal authentication flow.
+     *
+     * For most integrations, this method should be invoked in the onResume method of the Activity
+     * used to invoke
+     * [PayPalLauncher.launch].
+     *
+     * If the Activity used to launch the PayPal flow has is configured with
+     * android:launchMode="singleTop", this method should be invoked in the onNewIntent method of
+     * the Activity.
+     *
+     * @param pendingRequest the [PayPalPendingRequest.Started] stored after successfully
+     * invoking [PayPalLauncher.launch]
+     * @param intent         the intent to return to your application containing a deep link result
+     * from the PayPal browser flow
+     * @return a [PayPalPaymentAuthResult] that should be passed to
+     * [PayPalClient.tokenize] to complete
+     * the PayPal payment flow. Returns null if the user closed the browser to cancel the payment
+     * flow, or returned to the app without completing the PayPal authentication flow.
+     */
+    fun handleReturnToAppFromBrowser(
+        pendingRequest: PayPalPendingRequest.Started, intent: Intent
```

Comment:
Nit - not sure if we have a specific Kotlin style guide or if detekt will enforce these rule sets. What are your thoughts on having each parameter on a separate line when the primary constructor needs to expand more than a single line?

https://kotlinlang.org/docs/coding-conventions.html#class-headers

Reply:
Manually fixed this one but to enforce it across the project I think we need to turn on ArgumentListWrapping in the detekt config (and will add a bunch of unrelated files to this PR to fix). Maybe a separate PR would be a good time to evaluate the entire Kotlin style guide that we want to follow and update any rules. We don't currently have a strong Kotlin convention in this project so am open to any ruleset you all prefer.

Reply:
Sounds good. Yep definitely in a separate PR. 

We could add a baseline file to prevent all existing code from breaking the new ruleset. It would only enforce new code or modified code.

As for coding styles, I think Detekt's default ruleset follows the kotlin and Android style guides:
* https://developer.android.com/kotlin/style-guide
* https://kotlinlang.org/docs/coding-conventions.html

but we can also modify them if we need